core/templates/dev/head/pages/exploration_editor/simple_editor_tab/ScrollSyncService.js

```diff
@@ -0,0 +1,110 @@
+ */
+
+oppia.factory('ScrollSyncService', [
+  '$anchorScroll', '$document', '$rootScope', 'QuestionHashService',
+  'SimpleEditorManagerService',
+  function($anchorScroll, $document, $rootScope, QuestionHashService,
+           SimpleEditorManagerService) {
+    var targets = [];
+    var questionList = SimpleEditorManagerService.getQuestionList();
+
+    var clearHighlighted = function() {
+      angular.element('.highlighted').removeClass('highlighted');
+    };
+
+    var collapse = function(hash) {
+      var questions = questionList.getQuestions();
+      for (var i = 0; i < questions.length; i++) {
+        if (QuestionHashService.getQuestionHash(questions[i]) === hash) {
+          $rootScope.$broadcast(
+            'SimpleEditorSidebarToggleCollapse', questions[i]
+          );
+          break;
+        }
+      };
+    };
+
+    $anchorScroll.yOffset = 80;
+
+    var scrollSyncService = {
+      // Add an element to the targets list, when a user scrolls to an element
+      // in the target list, the sidebar is updated accordingly.
+      addTarget: function(target) {
+        targets.push(target);
+      },
+      // Update the active nav link
+      updateActive: function(hash) {
+        if (!angular.isDefined(hash)) {
+          return;
+        } else {
+          clearHighlighted();
+          var elm = angular.element('#sidebaritem-' + hash);
+
+          if (!(hash === 'intro' || hash === 'title')) {
+            elm.addClass('highlighted');
+          };
+
+          // Uncollapse the nav link in case hidden
+          if (elm.is(':hidden')) {
+            collapse(QuestionHashService.getParentQuestionHash(hash));
+          };
+          // TODO(andromfins): If the nav link is not displayed when the sidebar
+          // is overflown, scroll to it.
+        }
+      },
+      // Scroll to an element in the editor given its hash.
+      scrollTo: function(hash) {
+        $anchorScroll(hash);
+      }
+    };
+
+    // Onscroll event handler.
+    $document.on('scroll', function() {
+      // Find the closest positive target (the closest element in the target
+      // list to the top of the editor, with a positive distance from it).
+      var minPositiveTarget;
+      var documentY = $document.scrollTop();
+      for (var i = 0, foundPositive = false, minDy; i < targets.length; i++) {
+        var dy = targets[i].offset().top - documentY;
+        if (dy > 0) {
+          if (!foundPositive) {
+            minDy = dy;
+            minPositiveTarget = targets[i];
+            foundPositive = true;
+            continue;
+          } else {
+            if (dy < minDy) {
+              minDy = dy;
+              minPositiveTarget = targets[i];
+            }
+          }
+        }
+      };
+
+      if (!angular.isDefined(minPositiveTarget)) {
+        return;
+      } else {
+        // Update the sidebar active element.
+        scrollSyncService.updateActive(minPositiveTarget.attr('id'));
```

Comment:
I'm a bit confused, because here we're passing an id to updateActive(), but updateActive() takes a hash. What is the exact connection between id and hash?

If they're the same thing, let's just use "id" throughout. However, if they're different things, I recommend adding some documentation explaining what is what, otherwise things are going to get confusing.